### ISSUE TITLE
connection getting closed 

### ISSUE THREAD

**Issue by abhishekparwal** [brettwooldridge/HikariCP#582]
stacktrace ->
```
11:48:51.357 [ProvisioningClusteredScheduler_Worker-1] WARN  c.z.hikari.proxy.ConnectionProxy - iam--provisioning - Connection org.postgresql.jdbc42.Jdbc42Connection@6c0c4a6e, created Feb 16, 07:17:49.967, last release 16261389ms ago, IN_USE marked as broken because of SQLSTATE(08003), ErrorCode(0)
org.postgresql.util.PSQLException: This connection has been closed.
        at org.postgresql.jdbc2.AbstractJdbc2Connection.checkClosed(AbstractJdbc2Connection.java:853) ~[lib.jar:1.0-SNAPSHOT]
        at org.postgresql.jdbc3.AbstractJdbc3Connection.prepareStatement(AbstractJdbc3Connection.java:281) ~[lib.jar:1.0-SNAPSHOT]
        at org.postgresql.jdbc2.AbstractJdbc2Connection.prepareStatement(AbstractJdbc2Connection.java:326) ~[lib.jar:1.0-SNAPSHOT]
        at com.zaxxer.hikari.proxy.ConnectionProxy.prepareStatement(ConnectionProxy.java:265) ~[lib.jar:1.0-SNAPSHOT]
        at com.zaxxer.hikari.proxy.HikariConnectionProxy.prepareStatement(HikariConnectionProxy.java) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$5.doPrepare(StatementPreparerImpl.java:146) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$StatementPreparationTemplate.prepareStatement(StatementPreparerImpl.java:172) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.prepareQueryStatement(StatementPreparerImpl.java:148) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.loader.Loader.prepareQueryStatement(Loader.java:1928) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:1897) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:1875) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.loader.Loader.doQuery(Loader.java:919) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:336) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.loader.Loader.doList(Loader.java:2611) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.loader.Loader.doList(Loader.java:2594) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2423) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.loader.Loader.list(Loader.java:2418) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.loader.hql.QueryLoader.list(QueryLoader.java:501) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.hql.internal.ast.QueryTranslatorImpl.list(QueryTranslatorImpl.java:371) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.engine.query.spi.HQLQueryPlan.performList(HQLQueryPlan.java:220) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.internal.SessionImpl.list(SessionImpl.java:1268) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.internal.QueryImpl.list(QueryImpl.java:87) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.jpa.internal.QueryImpl.list(QueryImpl.java:567) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.jpa.internal.QueryImpl.getResultList(QueryImpl.java:436) [lib.jar:1.0-SNAPSHOT]
        at org.hibernate.jpa.criteria.compile.CriteriaQueryTypeQueryAdapter.getResultList(CriteriaQueryTypeQueryAdapter.java:50) [lib.jar:1.0-SNAPSHOT]
        at com.xyz.nmon.provisioning.jobs.amsaccount.dao.AmsAccountDao.readItem(AmsAccountDao.java:41) [lib.jar:1.0-SNAPSHOT]
        at com.xyz.nmon.provisioning.jobs.amsaccount.reader.AmsAccountSource.readItem(AmsAccountSource.java:25) [lib.jar:1.0-SNAPSHOT]
        at com.xyz.nmon.provisioning.jobs.amsaccount.job.AmsAccountProducerJob.lambda$execute$29(AmsAccountProducerJob.java:91) [lib.jar:1.0-SNAPSHOT]
        at com.xyz.nmon.provisioning.jobs.amsaccount.job.AmsAccountProducerJob$$Lambda$17/1816821542.call(Unknown Source) [lib.jar:1.0-SNAPSHOT]
        at rx.internal.operators.OperatorMap$1.onNext(OperatorMap.java:55) [lib.jar:1.0-SNAPSHOT]
        at rx.internal.util.ScalarSynchronousObservable$1.call(ScalarSynchronousObservable.java:43) [lib.jar:1.0-SNAPSHOT]
        at rx.internal.util.ScalarSynchronousObservable$1.call(ScalarSynchronousObservable.java:32) [lib.jar:1.0-SNAPSHOT]
        at rx.Observable$1.call(Observable.java:144) [lib.jar:1.0-SNAPSHOT]
        at rx.Observable$1.call(Observable.java:136) [lib.jar:1.0-SNAPSHOT]
        at rx.Observable$1.call(Observable.java:144) [lib.jar:1.0-SNAPSHOT]
        at rx.Observable$1.call(Observable.java:136) [lib.jar:1.0-SNAPSHOT]
        at rx.Observable$1.call(Observable.java:144) [lib.jar:1.0-SNAPSHOT]
        at rx.Observable$1.call(Observable.java:136) [lib.jar:1.0-SNAPSHOT]
        at rx.Observable$1.call(Observable.java:144) [lib.jar:1.0-SNAPSHOT]
        at rx.Observable$1.call(Observable.java:136) [lib.jar:1.0-SNAPSHOT]
        at rx.Observable.subscribe(Observable.java:7597) [lib.jar:1.0-SNAPSHOT]
        at rx.Observable.subscribe(Observable.java:7373) [lib.jar:1.0-SNAPSHOT]
        at com.xyz.nmon.provisioning.jobs.amsaccount.job.AmsAccountProducerJob.execute(AmsAccountProducerJob.java:104) [lib.jar:1.0-SNAPSHOT]
        at com.xyz.nmon.provisioning.jobs.amsaccount.job.AmsAccountProducerJob$$EnhancerByGuice$$1ff991c8.CGLIB$execute$0(<generated>) [lib.jar:1.0-SNAPSHOT]
        at com.xyz.nmon.provisioning.jobs.amsaccount.job.AmsAccountProducerJob$$EnhancerByGuice$$1ff991c8$$FastClassByGuice$$40006e9b.invoke(<generated>) [lib.jar:1.0-SNAPSHOT]
        at com.google.inject.internal.cglib.proxy.$MethodProxy.invokeSuper(MethodProxy.java:228) [lib.jar:1.0-SNAPSHOT]
        at com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed(InterceptorStackCallback.java:75) [lib.jar:1.0-SNAPSHOT]
        at com.xyz.indis.metrics.listener.TimedInterceptor.invoke(TimedInterceptor.java:56) [lib.jar:1.0-SNAPSHOT]
        at com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed(InterceptorStackCallback.java:75) [lib.jar:1.0-SNAPSHOT]
        at com.palominolabs.metrics.guice.TimedInterceptor.invoke(TimedInterceptor.java:26) [lib.jar:1.0-SNAPSHOT]
        at com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed(InterceptorStackCallback.java:75) [lib.jar:1.0-SNAPSHOT]
        at com.google.inject.internal.InterceptorStackCallback.intercept(InterceptorStackCallback.java:55) [lib.jar:1.0-SNAPSHOT]
        at com.xyz.nmon.provisioning.jobs.amsaccount.job.AmsAccountProducerJob$$EnhancerByGuice$$1ff991c8.execute(<generated>) [lib.jar:1.0-SNAPSHOT]
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202) [lib.jar:1.0-SNAPSHOT]
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573) [lib.jar:1.0-SNAPSHOT]
```

relevant dependencies ->

```
            <dependency>
                <groupId>org.hibernate</groupId>
                <artifactId>hibernate-core</artifactId>
                <version>5.0.1.Final</version>
            </dependency>


            <dependency>
                <groupId>com.zaxxer</groupId>
                <artifactId>HikariCP</artifactId>
                <version>2.4.1</version>
            </dependency>


            <dependency>
                <groupId>org.postgresql</groupId>
                <artifactId>postgresql</artifactId>
                <version>9.4-1205-jdbc42</version>
            </dependency>
```

Why hikari is closing the connections?

**Comment by brettwooldridge** [comment]
@abhishekparwal The error log indicates that the connection was "last released" 4.5 hours before this error, and created ~6.5 hours before.  There is no easy way to tell how long the connection was out of the pool before this error occurred, but it seems probably that it was out of the pool for a long time.  HikariCP tests the connection when it is borrowed from the pool, but once it is out of the pool there is nothing it can do to "keep it alive".

I recommend turning on leak detection, by setting the ``leakDetectionThreshold``.  If you set the threshold to 5 or 10 minutes (600000ms), you might find that this connection is being held out of the pool for too long.

Even settings like ``maxLifetime`` (which defaults to 30 minutes), will not be effective for connection that are borrowed from the pool, because HikariCP will never interrupt an "in-use" connection.

**Comment by abhishekparwal** [comment]
After this error comes,  Hikari is not able to create a connection again.
